File: apps/api-go/internal/links/redirect.go
package links

import (
	"errors"
	"net/http"

	"github.com/Lisovate/acuvis-demo/apps/api-go/internal/analytics"
	"github.com/Lisovate/acuvis-demo/apps/api-go/internal/passwords"
	"github.com/go-chi/chi/v5"
)

// RedirectHandler returns a handler that does the slug → target redirect,
// enforcing expiration and password gates, and recording a click before
// the 302. Clicks are written inline; production should swap this for an
// async queue.
func RedirectHandler(linkSvc *Service, clickSvc *analytics.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		slug := chi.URLParam(r, "slug")
		link, err := linkSvc.FindBySlug(slug)
		if err != nil {
			if errors.Is(err, ErrNotFound) {
				writeError(w, http.StatusNotFound, "link not found")
				return
			}
			writeError(w, http.StatusInternalServerError, "lookup failed")
			return
		}

		// Expiration check. Null expires_at → never expires.
		if link.Expired() {
			writeError(w, http.StatusGone, "link expired")
			return
		}

		// Password gate.
		if link.HasPassword() {
			supplied := r.URL.Query().Get("password")
			if supplied == "" {
				writeError(w, http.StatusUnauthorized, "password required")
				return
			}
			if !passwords.Verify(supplied, link.PasswordHash) {
				writeError(w, http.StatusForbidden, "invalid password")
				return
			}
		}

		// Record the click before issuing the redirect — analytics are
		// eventually-consistent but writing inline keeps the demo simple.
		_ = clickSvc.Record(link.ID, r.RemoteAddr, r.Header.Get("User-Agent"), r.Header.Get("Referer"))

		http.Redirect(w, r, link.TargetURL, http.StatusFound)
	}
}

File: apps/api-python/app/routes/redirect.py
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from fastapi.responses import RedirectResponse
from sqlalchemy.orm import Session

from ..db import get_session
from ..lib.clicks import record_click
from ..lib.passwords import verify_link_password
from ..models import Link

router = APIRouter(tags=["redirect"])


@router.get("/{slug}", include_in_schema=False)
def redirect(
    slug: str,
    request: Request,
    session: Annotated[Session, Depends(get_session)],
    password: Annotated[str | None, Query()] = None,
) -> RedirectResponse:
    link = session.query(Link).filter(Link.slug == slug).first()
    if link is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="link not found")

    # Expiration check. Links with NULL expires_at never expire.
    if link.expires_at is not None and link.expires_at < datetime.utcnow():
        raise HTTPException(status.HTTP_410_GONE, detail="link expired")

    # Password gate.
    if link.password_hash is not None:
        if password is None:
            raise HTTPException(status.HTTP_401_UNAUTHORIZED, detail="password required")
        if not verify_link_password(password, link.password_hash):
            raise HTTPException(status.HTTP_403_FORBIDDEN, detail="invalid password")

    # Record the click before issuing the redirect — analytics are
    # eventually-consistent but writing inline keeps the demo simple.
    record_click(
        session=session,
        link_id=link.id,
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
        referer=request.headers.get("referer"),
    )

    return RedirectResponse(url=link.target_url, status_code=status.HTTP_302_FOUND)
